### ISSUE TITLE
Recycling is strange

### ISSUE THREAD

**Issue by alexroeber** [notfood/RimWorld-MendAndRecycle#6]
So, I found recycling to be rather strange and thought I'd read the code to understand my confusion. Well, I'm more confused now.
1. The WorkAmount of RecycleApparel is set to 100, which is solely used in line 32 of JobDriver_Recycle which ignores any value >10.
2. Line 71 of JobDriver_Recycle calculates the skillFactor for the SuccessChanceUtil.
float skillFactor = Mathf.Lerp (0.5f, 1.5f, skillPerc); <=> float skillFactor = skillPerc + 0.5f
Why use Mathf.Lerp for that? Either way, lv0->0.5, lv10->1, lv20->1.5.
But that's not the important part:
3. SuccessChanceUtil now uses this skillFactor (which is now named skillFactory, very interesting) to calculates this success chance, by this formula:
return random.Next() % 101 >= baseFc * skillFactor
baseFc is some fail chance, even changeable by the user, I'll use 1 so I don't need to calculate anything.
Just to be sure, what this formula does is take a random number between 0 and 100 and test if it's greater than or equal our skillFactor. Our success rate with skill level 0 is therefore 99.5%(100-0.5) and 98.5%(100-1.5) with skill level 20.
I conclude it's harder to recycle the better you get.
4. After reading the code, I also understand my initial confusion and the reason for this all. Faster working pawns are not recycling any faster.

My advice:
You're using Mathf.Lerp anyway, so simply swap 0.5 and 1.5 to Mathf.Lerp (1.5f, 0.5f, skillPerc).
And how about multiplying fixedHitPointsPerCycle by WorkSpeedGlobal or TailoringSpeed or something? Just an idea.

Sorry for all this text, but I love this mod, so I mind mistakes.

**Comment by notfood** [comment]
Mm, so the bug report is, there is no way to speed it up?

It uses StatDefOf.WorkToMake of the pawn which is affected by WorkSpeedGlobal if I'm not mistaken. A fast pawn will finish the cycle faster. Get their WorkToMake stat high and try it.

Success rate is actually the chance of failing and slashing 50hp to the object and getting nothing out of it. It happens often at low levels and rarely happens at max level.

1. Workamount is defined by the def, it sets the amount a pawn requires per cycle.
2. Lerp is fine, what do you suggest?
3. The better the pawn the less chance to ruin it.
4. It uses the WorkToMake stat which is goberned by crafting+handling.

**Comment by alexroeber** [comment]
Okay, I'm pretty sure my math was correct, so would be my conclusion.

**Comment by alexroeber** [comment]
Whoops, shouldn't press Ctrl+Enter
1. Yeah, no, that's the problem.
2. Lerp is useless in that case, but if you get to
3. you can use it just fine, just swap the first 2 parameters. My conclusion should be correct that's why I calculated the success chance for baseFc = 1.
4. Ugh god, some semi-extensive research later.... This... `StatExtension.GetStatValue (pawn, StatDefOf.WorkToMake, true);`
It seems to be important. I poured over the whole decompiled code and the Def of WorkToMake and still have no idea why your recycling even works... As far as I know, it should return 0, simply because of `<category>BasicsNonPawn</category>`. No pawn should have that stat and there is no post-processing for WorkToMake to change that. Although StatWorker.FinalizeValue does some strange things, dunno.

Conclusion: Seriously your speed management is all over the place, the more I learn, the worse it gets. Also: your 4. is simply wrong.

Suggestion:
1. Swap those parameters in Lerp, I calculated it for a reason, I'm pretty sure correctly, so it's wrong as it is now.
2. workCycleProgress = workCycle = [something] is pretty solid but that [something] is all wrong. If you want to put it in the Def, which I suggest, use that number, put it to 10 if you want to but leave it there.
3. workCycleProgress -= [something] is also damn obvious, but its [something] is way worse. WorkToMake is pretty much the equivalent of workAmount in case you don't have a dedicated RecipeDef but simply a RecipeMaker inside a ThingDef. There is also no single CreateStuffSpeed or whatever you want there, but since you're recycling at a tailor bench, I'd use TailorSpeed. Or you could create your very own RecycleSpeed by copying/adjusting TailorSpeed which would then correctly be affected by WorkSpeedGlobal. Which is something you want as I gather from your text.

Sorry, this is not a "normal" bug report. I can read code, that's why I don't have to hope that you find any mistakes yourself and you don't need to use your free time for debugging.

**Comment by notfood** [comment]
1.
Check these lines
https://github.com/notfood/RimWorld-MendAndRecycle/blob/master/Source/JobDriver_Recycle.cs#L73
https://github.com/notfood/RimWorld-MendAndRecycle/blob/master/Source/FailChanceByQuality.cs#L58
`SuccessChanceUtil.SuccessOnAction` expects a positive range
Let's set baseFc to 0.5 just for fun.
Try with 1.5 => 0.5 * 1.5
results in random >= 0.75

Now try with 0.5 => 0.5 * 0.5
results in random >= 0.25

With your suggestion, a pawn with 0 skill, from the Lerp would be getting 1.5 factor, combined with base factor of 0.5, they'd be getting a chance of 0.75 to not damage it while a pawn with 20 skill would have a 0.25 chance of not damaging it.

2.
3.
You made me have some doubts about WorkToMake, I'll check it again. I referenced the sculpture making job driver, they're very similar. The logic behind it is solid. It substracts some work the pawn can do per cycle.

**Comment by alexroeber** [comment]
Uhm... Skillllevel 0 means skillPerc := 0 => Lerp gets you 0.5, Skilllevel 20 means skillPerc := 1=> Lerp gets you 1.5. The lines I should check have nothing to do with that! Your calculations are correct, your conclusion is not, you mixed them up.
And I know that your logic is solid, I even said that myself, the implementation is still wrong.

**Comment by alexroeber** [comment]
Since nothing happened, I downloaded the mod and will fix it myself, without publishing it of course. I'll just use my own version...
Funny, looking at mending, I discovered that it has the same problem with the speed, but the probability  bug does not exist, simply because of a correct usage of Lerp....
Whatever, I will close this now, since I don't care anymore.

**Comment by notfood** [comment]
Reworked it, would appreciate your input.

**Comment by alexroeber** [comment]
Well... I saw your comment and thought I'd at least honor your wish, but as far as I can see, you did not change anything related to this issue. You know, I copied this repository, fixed every relevant aspect and changed a few parameters to make it "natural", so mostly to match normal tailoring. I am satisfied enough with my personal version to seriously not care about your mod anymore. If you really want my help, start by understanding my very first comment, there is sufficient input.
